internal/venice-consumer/src/main/java/com/linkedin/venice/client/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -0,0 +1,176 @@
+
+  public VeniceAfterImageConsumerImpl(
+      ChangelogClientConfig changelogClientConfig,
+      Consumer<KafkaKey, KafkaMessageEnvelope> kafkaConsumer) {
+    super(changelogClientConfig, kafkaConsumer);
+    this.currentTopic = Version.composeKafkaTopic(storeName, storeCurrentVersion);
+  }
+
+  public VeniceAfterImageConsumerImpl(ChangelogClientConfig changelogClientConfig) {
+    this(changelogClientConfig, new KafkaConsumer<>(changelogClientConfig.getConsumerProperties()));
+  }
+
+  @Override
+  public Collection<PubSubMessage<K, ChangeEvent<V>, Long>> poll(long timeoutInMs) {
+    List<PubSubMessage<K, ChangeEvent<V>, Long>> pubSubMessages = new ArrayList<>();
+    ConsumerRecords<KafkaKey, KafkaMessageEnvelope> consumerRecords = kafkaConsumer.poll(timeoutInMs);
+    for (ConsumerRecord<KafkaKey, KafkaMessageEnvelope> consumerRecord: consumerRecords) {
+      PubSubTopicPartition pubSubTopicPartition = getPubSubTopicPartitionFromConsumerRecord(consumerRecord);
+      long offset = consumerRecord.offset();
+      int payloadSize = consumerRecord.serializedKeySize() + consumerRecord.serializedValueSize();
+      long timestamp = consumerRecord.timestamp();
+      if (consumerRecord.key().isControlMessage()) {
+        ControlMessage controlMessage = (ControlMessage) consumerRecord.value().payloadUnion;
+        if (handleControlMessage(controlMessage, pubSubTopicPartition)) {
+          return pubSubMessages;
+        }
+      } else {
+        byte[] keyBytes = consumerRecord.key().getKey();
+        K currentKey = keyDeserializer.deserialize(keyBytes);
+        if (isReadFromChangeCaptureTopic) {
+          Optional<PubSubMessage<K, ChangeEvent<V>, Long>> pubSubMessage =
+              convertConsumerRecordToPubSubChangeEventMessage(consumerRecord, pubSubTopicPartition);
+          if (pubSubMessage.isPresent()) {
+            pubSubMessages.add(
+                new ImmutablePubSubMessage<>(
+                    pubSubMessage.get().getKey(),
+                    new ChangeEvent<>(null, pubSubMessage.get().getValue().getCurrentValue()),
+                    pubSubTopicPartition,
+                    offset,
+                    timestamp,
+                    keyBytes.length + currentValuePayloadSize[pubSubTopicPartition.getPartitionNumber()]));
+          }
+        } else {
+          PubSubMessage<K, ChangeEvent<V>, Long> pubSubMessage = convertRecordToPubSubMessage(
+              consumerRecord.value(),
+              currentKey,
+              pubSubTopicPartition,
+              offset,
+              timestamp,
+              payloadSize);
+          pubSubMessages.add(pubSubMessage);
+        }
+      }
+    }
+    return pubSubMessages;
+  }
+
+  private boolean handleControlMessage(ControlMessage controlMessage, PubSubTopicPartition pubSubTopicPartition) {
+    ControlMessageType controlMessageType = ControlMessageType.valueOf(controlMessage);
+    if (controlMessageType.equals(ControlMessageType.END_OF_PUSH)) {
```

Comment:
I think we'll need to revisit this logic at some point so lets at a TODO.  As I think on it, seeking up to EOP isn't a great solution.  It is problematic in the following scenario:

We have key1=∆ in the original version push.  After some time (after EOP) we update the key to be key1=∫ and then update it again to be key1=µ.  Then compaction runs on both the change capture topic.  The change capture topic will have an event that looks like:

{ beforeImage: key1=∫ AfterImage: key1=µ} however, in the version topic if we consumed up to EOP (and compaction didn't run on the VT) we'll have consumed key1=∆.  From the user perspective, we skipped the delta between ∆->∫.  I think initially this is ok, it's a given that consumption of boostrap data is a special kind of case, but I'm pretty sure we can do better.  

If we're able to consume on the VT past EOP and consume up to a fairly recent event (preferably up to data which comes after compaction in both the VT and change capture topic) then we can be sure we won't have any missed history.  I'm not sure how queryable that information is from the client, but I think it'd be good enough to either a)consume on VT up to some lag threshhold, kind of like how we do ready to serve checks, where the lag threshhold is some value which is smaller then the frequency by which kafka runs compaction or b) consume on VT up to the end of VT or c) consume on VT up to the point when the client started consuming on VT in the first place (this one isn't perfect, but it's probably close enough and might be easier to code?)

Reply:
Thanks for the comment! I think the problem is about the compaction causing data gap between VT and change capture VT. So the solution you proposed here is to read as much as we can on the VT before we switch to change capture topic for this after image use case (Actually I did not quite understand of solution 'c' you proposed). I feel that our current thinking is relying on the data from change capture topic.  So the question is: is it possible that we do not rely on the data from change capture topic, but only the version swap message from change capture topic to guide the client that the newer version is good to read (we might need to have some logic periodically check the very end of the change capture topic to find the version swap is there or not), then we can start to consume newer VT from the very beginning and filtering the events we have seen.